LwM2M: Zephyr v2.0.0 changes (new objects + bug fixes + enhancements)

mike-scott opened the pull request:
This patch series adds the following new LwM2M objects:
- Connectivity Monitoring (ID 4): This LwM2M Object enables monitoring of parameters related to network connectivity.
- Location (ID 6): This LwM2M object provides a range of location telemetry related information.
- IPSO Accelerometer: This IPSO object can be used to represent a 1-3 axis accelerometer.
- IPSO Push Button: This IPSO object is used to report the state of a momentary action push button control and to count the number of times the control has been operated since the last observation.
- IPSO On/Off Switch: This IPSO object should be used with an On/Off switch to report the state of the switch.
- IPSO Buzzer: This IPSO object should be used to actuate an audible alarm such as a buzzer, beeper, or vibration alarm.

Bug fixes:
- Add missing "App type" resource to 
- Fix string handling for resource instances

Feature completions / Enhancements:
- Reworked how resource instances are handled: now generic functions can access multi-instance resources by providing the 4th portion of the LwM2M path Ex: "4/0/1/0" for the first available bearer and "4/0/1/1" for the 2nd available bearer, etc.
- Removed an unused field in the object field data
- Use server record to set default observe notify timing instead of hard-coded values
- Remove data storage for IPSO Timer application type (can be set by user)

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+static s8_t avail_net_bearer[CONNMON_AVAIL_BEARER_MAX];
+static s8_t rss;
+static u8_t link_quality;
+static char ip_address[CONNMON_IP_ADDRESS_MAX][INET6_ADDRSTRLEN];
```
These IP address strings use quite a lot of memory as the IPv6 string is 46 bytes long. Would it make sense to store the binary value instead and just convert to/from string when needed?

mike-scott replied:
My original patches used `struct net_addr` for storage.  I can revert back to that or some equivalent.  In my head implementing the IPs as strings seemed easier for the user, so I changed.

It also exposed a much deeper issue I'm having with resource instances atm.  It's just too hard for the user to add / delete these instances in any sane manner.  Each of these multi-instance resource is far too hard to work with at the moment.  I have another patch set I'm working on to make creating / deleting and setting the multi-instance resources basically the same as any other resource.

Here's an example from the lwm2m sample where we set the data buffer for the `Manufacturer` field:
```
	lwm2m_engine_set_res_data("3/0/0", CLIENT_MANUFACTURER,
				  sizeof(CLIENT_MANUFACTURER),
				  LWM2M_RES_DATA_FLAG_RO);
```
We access the resource via the established "3/0/0" URI (object id, object instance id and resource id)

For us to set the first 2 instances of the Connectivity Monitoring IP addresses we *should* be able to use something like:
```
	lwm2m_engine_create_res_inst("4/0/4/0");
	lwm2m_engine_set_res_data("4/0/4/0", const_string, length of cont_string,
				  LWM2M_RES_DATA_FLAG_RO);
	lwm2m_engine_create_res_inst("4/0/4/1");
	lwm2m_engine_set_res_data("4/0/4/1", const_string2, length of cont_string2,
				  LWM2M_RES_DATA_FLAG_RO);
```
Not only does this remove the need for pre-allocating the space for these IP addresses, but also allows for non-sequential resource instance id's etc.  (It's up to the user).   The user could also setup a pre-read callback where it takes the current IP address and converts it to a string on the fly and hands it back whenever it's read.  Etc.  Endless more opportunities.

Instead, because of the way the resource instances are implemented atm, we use these crazy 1-off functions like:
`lwm2m_connmon_add_bearer()`
`lwm2m_connmon_remove_bearer()`
`lwm2m_connmon_add_ip_address()`
and we have to pre-allocate all of the resource data during the object instance creation.

jukkar replied:
> Not only does this remove the need for pre-allocating the space for these IP addresses, but also allows for non-sequential resource instance id's etc. (It's up to the user).

Indeed this sounds much better solution. We can leave the strings for now if in the future we do not really need them anyway.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+{
+	int index;
+
+	if (bearer_id >= 50) {
```
It would be more clearer if we check against a define instead of magic value, now it is very unclear what this special value means.

mike-scott replied:
Most of the odd-ball logic was removed from the connectivity monitoring object, so this issue goes away.  It will need to be re-addressed if we later integrate with the networking subsystem somehow.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+		return -EINVAL;
+	}
+
+	for (index = 0; index < CONNMON_AVAIL_BEARER_MAX; index++) {
```
Please check using `index < ARRAY_SIZE(avail_net_bearer)` so that upper limit is always correct.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+ */
+int lwm2m_connmon_remove_bearer(int index)
+{
+	if (index < 0 || index >= CONNMON_AVAIL_BEARER_MAX) {
```
check against `ARRAY_SIZE(avail_net_bearer)` instead

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+	return index;
+}
+
+int lwm2m_connmon_add_ip_address(char *addr)
```
Perhaps `const char *` here?

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+	return index;
+}
+
+int lwm2m_connmon_add_router_ip_address(char *addr)
```
ditto

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+	return index;
+}
+
+int lwm2m_connmon_add_apn(char *new_apn)
```
here too

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+	link_quality = 0U;
+	mnc = 0U;
+	mcc = 0U;
+	for (i = 0; i < CONNMON_AVAIL_BEARER_MAX; i++) {
```
empty line before the `for`

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+	for (i = 0; i < CONNMON_AVAIL_BEARER_MAX; i++) {
+		avail_net_bearer[i] = -1;
+	}
+	for (i = 0; i < CONNMON_APN_MAX; i++) {
```
same here

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+	for (i = 0; i < CONNMON_APN_MAX; i++) {
+		apn[i][0] = '\0';
+	}
+	for (i = 0; i < CONNMON_IP_ADDRESS_MAX; i++) {
```
here

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,397 @@
+	for (i = 0; i < CONNMON_IP_ADDRESS_MAX; i++) {
+		ip_address[i][0] = '\0';
+	}
+	for (i = 0; i < CONNMON_ROUTER_IP_ADDRESS_MAX; i++) {
```
and here too

jukkar commented on include/net/lwm2m.h:
```diff
@@ -267,6 +267,93 @@ int lwm2m_device_set_pwrsrc_current_ma(int index, int current_ma);
+ * @brief Register an available net bearer with the LwM2M Connectivity
+ *        Monitoring object.
+ *
+ * @param[in] bearer_id Network Bearer ID.
```
What exactly is the bearer_id value, is this related to network interface or something that exists already in the network stack?

mike-scott replied:
This was removed in the latest patch series as it is now up to the user to set these resources via generic methods.  But here's the description of what a "bearer" is according to the Connectivity Monitoring spec:
```
Indicates the network bearer used for the current LwM2M communication session
from the network bearer list below. 
 The number range is split into three categories:
 0 - 20 are Cellular Bearers 
 21 - 40 are Wireless Bearers 
 41 - 50 are Wireline Bearers
 
 More specifically: 
 0: GSM cellular network
 1: TD-SCDMA cellular network
 2: WCDMA cellular network
 3: CDMA2000 cellular network
 4: WiMAX cellular network
 5: LTE-TDD cellular network
 6: LTE-FDD cellular network
 7: NB-IoT
 8 - 20: Reserved for other types of cellular network
 21: WLAN network
 22: Bluetooth network
 23: IEEE 802.15.4 network
 24 - 40: Reserved for other types of local wireless network
 41: Ethernet
 42: DSL
 43: PLC
 44 - 50: reserved for other types of wireline networks.
```

jukkar commented on subsys/net/lib/lwm2m/ipso_onoff_switch.c:
```diff
@@ -0,0 +1,222 @@
+{
+	int i, ret = -ENOENT;
+
+	for (i = 0; i < MAX_INSTANCE_COUNT; i++) {
```
Better to use `ARRAY_SIZE(inst)`

jukkar commented on subsys/net/lib/lwm2m/ipso_onoff_switch.c:
```diff
@@ -0,0 +1,222 @@
+	int index, avail = -1, i = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
```
same here

jukkar commented on subsys/net/lib/lwm2m/ipso_push_button.c:
```diff
@@ -0,0 +1,152 @@
+{
+	int i, ret = -ENOENT;
+
+	for (i = 0; i < MAX_INSTANCE_COUNT; i++) {
```
Better `ARRAY_SIZE(inst)`

jukkar commented on subsys/net/lib/lwm2m/ipso_push_button.c:
```diff
@@ -0,0 +1,152 @@
+	int index, avail = -1, i = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
```
here too

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1956,11 +1956,11 @@ static int lwm2m_read_handler(struct lwm2m_engine_obj_inst *obj_inst,
-					  strlen((u8_t *)data_ptr));
+					  ((u8_t *)data_ptr + i*data_len),
+					  strlen(((u8_t *)data_ptr +
+							  i*data_len)));
```
spaces around `*` in two lines above

mike-scott replied:
This logic was removed in favor of each resource instance having it's own data_ptr (as it should be).

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_server.c:
```diff
@@ -98,6 +98,32 @@ static int update_trigger_cb(u16_t obj_inst_id)
+{
+	int i;
+
+	for (i = 0; i < MAX_INSTANCE_COUNT; i++) {
```
Better use `ARRAY_SIZE()` here

jukkar commented on subsys/net/lib/lwm2m/ipso_buzzer.c:
```diff
@@ -0,0 +1,248 @@
+{
+	int i, ret = -ENOENT;
+
+	for (i = 0; i < MAX_INSTANCE_COUNT; i++) {
```
Better use `ARRAY_SIZE()`

jukkar commented on subsys/net/lib/lwm2m/ipso_buzzer.c:
```diff
@@ -0,0 +1,248 @@
+	int index, avail = -1, i = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
```
ditto

mike-scott commented:
@jukkar thanks for the great review.  I'll clean this up.

mike-scott commented:
@jukkar After thinking about it.  I'm going to remove the new objects introduced in this PR.   Leaving only the fixes and finishing of functionality.

The next PR I put up will have resource instance management fixes, and then I'll submit the new objects last -- that way they will use the new resource instance changes.

mike-scott committed:
net: lwm2m: use server record to set default observe notify timing

Server records contain the default PMIN and PMAX settings for how
often we can send observe notifications.  We are currently using
arbitrary defaults which cannot be changed when compiled or
during runtime.

Let's add Kconfig settings for the default settings to use and
also lookup the current values in the active server record when
an observe is added.

The actual PMIN/PMAX values can still be set via WRITE_ATTRIBUTE
operation.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: remove data storage for IPSO Timer application type

We are pre-allocating a storage variable for the application type
resource in the IPSO Timer object.  This is an optional resource
which won't always be set by samples.

Let's leave out the pre-allocated variable and let the sample set
this reference if needed (it's optional).

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add missing application type to IPSO Light Control

Per IPSO Light Control definition from the OMA LwM2M registry:
http://www.openmobilealliance.org/tech/profiles/lwm2m/3311.xml

There is an optional "Application Type" string resource (5750) in the
Light Control object.  This was missed in the initial implementation.

NOTE: sample will assign reference if needed.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
samples: net: lwm2m_client: set LWM2M_COAP_BLOCK_SIZE to 512

The native Zephyr LwM2M client enables the JSON formatter which
can use more of the packet buffers than the default 256 with
TLV formatting.  Let's set the default to 512 in order to avoid
cutting off the output of larger READ operations.

In the future, we should establish block transfer buffers to
handler longer READ ops.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
samples: net: lwm2m_client: send temp value back to engine

By sending the temp value back to the LwM2M engine, it allows all of
the min/max values to be updated over time.

Fixes an issue where min/max values were not being updated.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: update function prototypes and descriptions

- Several of the functions use "path" as the parameter name for the
  string-based LwM2M path.  Let's clarify by using "pathstr".
- Recent updates to the LwM2M engine now support resource instances
  when parsing the LwM2M path.  Let's update descriptions accordingly.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: remove unused multi_max_count field from obj_field

The multi_max_count is no longer used and can be removed from the
obj_field structure if we change all of the OBJ_FIELD() macros to
use OBJ_FIELD_DATA() instead.

Technically, OBJ_FIELD() and OBJ_FIELD_DATA() are now the same, but
we're keeping them both for the time being.  In the future, more
fields may be added to the obj_field structure and we can use the
OBJ_FIELD() macro again if that's the case.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: remove unique lwm2m_engine_obj_delete_cb_t definition

Instead, let's use lwm2m_engine_user_cb_t which is used elsewhere and
matches the same signature.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: remove lwm2m_engine_obj from most handlers/formatter OPs

Due to work combining data into the lwm2m_message structure, we no
longer need to pass the lwm2m_engine_obj parameter between
formatters and most of the operation handlers.

So, let's remove it.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: rework resource instance storage / access methods

LwM2M allows for multiple instance resources such the power source
resources in the device object.  These types of resources have
always been very hard to work with, and frankly were poorly
implemented.

This led to other issues where it was very hard to have
non-sequential resource instances, and each resource of this type
needed special getter / setter methods such as:
lwm2m_device_add_pwrsrc()
lwm2m_device_set_pwrsrc_voltage_mv()

Going forward, as more LwM2M objects are implemented this just
doesn't scale well.

To fix this:
- split the resource instance data out from the resource data.
  This includes the data pointer information and resource
  instance id.
- add resource id and resource instance id to the event callback
  functions so user's can see in more detail what resources and
  resource instances are being handled.
- allow generic functions like lwm2m_engine_get_*() and
  lwm2m_engine_set_*() to access resource instance data.
- adjust object resource initialization  macros to map resource
  instances to resources at the time of object instance
  creation.
- fix up the lwm2m_client as a reflection of all of these changes.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add IPSO Buzzer object support

The IPSO Buzzer object is used to represent a buzzer, beeper or
vibrating alarm.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add IPSO On/Off Switch object support

This object is used with an On/Off switch to report it's state.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add IPSO Push Button object support

This Object is used to report the state of a momentary action push
button control and to count the number of times the control has
been operated since the last observation.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add Connection Monitoring object support

This core LwM2M Object enables monitoring of parameters related to
network connectivity.

This is only the basic object structure.  More work will be needed
to set the various resources based on connectivity.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add Location object support

This core LwM2M object provides a range of location telemetry related
information.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add IPSO Accelerometer object support

This IPSO object can be used to represent a 1-3 axis accelerometer.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
samples: net: lwm2m_client: support for compiling in new objects

Let's add support for recently added LwM2M objects:
- Connectitiy Monitoring
- Location
- IPSO Accelerometer
- IPSO Buzzer
- IPSO On/Off Switch
- IPSO Push Button

Currently, this is limited to compile enabled and manual creation
of objects via LwM2M server.  Objects will respond to read/write
operations, but not have real HW backing.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add LwM2M path to engine_set errors

When presenting errors in lwm2m_engine_set() let's include the related
LwM2M path for easier debugging.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
@jukkar I apologize for the size of this PR.  I *think* I addressed your comments and this now includes the work to fixup the resource instance handling as discussed.

mike-scott commented:
Hm.. looking at checkpatch issues.

jukkar commented on subsys/net/lib/lwm2m/ipso_buzzer.c:
```diff
@@ -0,0 +1,267 @@
+{
+	/* Set default values */
+	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
+	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
```
These two memset() seems to be useless as the variables are static.

mike-scott replied:
You are correct!  And this was copy/pasted into every single LwM2M object that is in Zephyr.  I'm going to do 2 things here:
1. Remove these memset lines from the *_init() functions of the new objects being introduced.
2. Add a memset line to the *_create() function above here to make sure the resource data that the new object instance is using is reset (it could have been used prior and then freed up again via DELETE op)
3. Stage a patch at the end of this series doing the same for all of the existing objects.

jukkar commented on subsys/net/lib/lwm2m/ipso_onoff_switch.c:
```diff
@@ -0,0 +1,245 @@
+{
+	/* Set default values */
+	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
+	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
```
Same here, useless memset()

mike-scott replied:
will be cleaned up in upcoming revision

jukkar commented on subsys/net/lib/lwm2m/ipso_push_button.c:
```diff
@@ -0,0 +1,168 @@
+{
+	/* Set default values */
+	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
+	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
```
here too

mike-scott replied:
will be cleaned up in upcoming revision

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_location.c:
```diff
@@ -0,0 +1,118 @@
+
+	/* Set default values */
+	(void)memset(&inst, 0, sizeof(inst));
+	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) * LOCATION_MAX_ID);
```
here too

mike-scott replied:
will be cleaned up in upcoming revision

jukkar commented on subsys/net/lib/lwm2m/ipso_accelerometer.c:
```diff
@@ -0,0 +1,143 @@
+{
+	/* Set default values */
+	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
+	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
```
here too

mike-scott replied:
will be cleaned up in upcoming revision

jukkar commented on samples/net/lwm2m_client/prj.conf:
```diff
@@ -29,6 +29,13 @@ CONFIG_LWM2M_IPSO_SUPPORT=y
 CONFIG_LWM2M_IPSO_TEMP_SENSOR=y
```
Not necessarily part of this PR, but I just noticed that we are missing lwm2m settings from `tests/net/all/prj.conf` file. The file tries to enable max number of options from networking subsystem. It would be great to have lwm2m there too. There is `tests/net/all/check_net_options.sh` that is currently missing lwm2m checks and can be used to generate a list of currently missing options from prj.conf file. As I said, no need to add this support in this PR (so that we get this one merged sooner than later) but it would be nice to have lwm2m in tests/net/all at some point.

mike-scott replied:
Referencing this issue so we can track this (which has been open for some time): https://github.com/zephyrproject-rtos/zephyr/issues/4634

jukkar reviewed (commented):
There was lot of code so I probably missed something. Just some clarifications needed + minor nit picks.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_connmon.c:
```diff
@@ -0,0 +1,173 @@
+
+#define CONNMON_STRING_SHORT			8
+
+#if defined(CONFIG_LWM2M_CONN_MON_BEARER_MAX)
```
Hmm, is it even possible to have this config undefined, if `LWM2M_CONN_MON_OBJ_SUPPORT` is enabled? If not, I'd remove this check to keep the code cleaner.

mike-scott replied:
Yep, it cannot be undefined.  Removing this check.  Thanks.

rlubos reviewed (commented):
That's a lot of code, looks pretty neat.
Nothing to complain about really, LGTM after Jukkas comments are addressed.

mike-scott commented on samples/net/lwm2m_client/src/lwm2m-client.c:
```diff
@@ -341,7 +343,8 @@ static int lwm2m_setup(void)
 	lwm2m_engine_register_post_write_callback("3340/0/5543",
 			timer_digital_state_cb);
-	lwm2m_engine_set_string("3340/0/5750", "Test timer");
+	lwm2m_engine_set_res_data("3340/0/5750", TIMER_NAME, sizeof(TIMER_NAME),
```
Great question.  Yes the full size is what we want here (even for strings).
When we make assignments to that buffer via lwm2m_engine_set_string() there is a check in the engine which accounts for the size - 1:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/lwm2m/lwm2m_engine.c#L1417

mike-scott commented on samples/net/lwm2m_client/src/lwm2m-client.c:
```diff
@@ -337,6 +338,9 @@ static int lwm2m_setup(void)
 		lwm2m_engine_register_post_write_callback("3311/0/5850",
 				led_on_off_cb);
+		lwm2m_engine_set_res_data("3311/0/5750",
+					  LIGHT_NAME, sizeof(LIGHT_NAME),
```
Also good here per above explanation.

mike-scott commented on samples/net/lwm2m_client/src/lwm2m-client.c:
```diff
@@ -46,6 +46,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define CLIENT_DEVICE_TYPE	"OMA-LWM2M Client"
```
Great catch!  Updating to the OMA LwM2M registry xml definition:
http://www.openmobilealliance.org/tech/profiles/lwm2m/3311.xml

mike-scott committed:
net: lwm2m: cleanup memset usage during init

Several problems with memset usage in the LwM2M subsystem were
identified:
- Every single object that can have multiple instances is using
  memset to initialize static resource data during init.  This data
  will already be set to 0 because it is static, so the memset
  statements are unneeded.
- Instead of using memset during object init which is only called
  one time during kernel startup, let's add a memset to the
  object create function to ensure the resource data is cleared out.
  It could have been used prior and then released via a DELETE op.
- the IPSO Timer object was setting a lot of data structure members
  to 0 in the create function.  Let's do 1 memset on the entire
  structure and then only the non-zero values afterward.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Pushed a new patchset addressing the comments above as well as some minor cleanup:
- Removed the range setting on BEARER_MAX and APN_MAX and default to 1 instance -- this felt like over-protection in general.
- In some of the objects, I got rid of multiple vars being set to 0 during create() where they belonged to a structure.  Instead just memset the whole structure first.
- In Connectivity Monitor I moved the instance defaults into create() just like every other object.

Thanks for sticking with me on this review.

For ease of review, here's the diff from the last version:
```
diff --git a/subsys/net/lib/lwm2m/Kconfig b/subsys/net/lib/lwm2m/Kconfig
index 711c5a3689..ab9dbc11af 100644
--- a/subsys/net/lib/lwm2m/Kconfig
+++ b/subsys/net/lib/lwm2m/Kconfig
@@ -159,8 +159,7 @@ config LWM2M_CONN_MON_OBJ_SUPPORT
 config LWM2M_CONN_MON_BEARER_MAX
 	int "Maximum # of available network bearer resource instances"
 	depends on LWM2M_CONN_MON_OBJ_SUPPORT
-	default 5
-	range 1 20
+	default 1
 	help
 	  This value sets the maximum number of available network bearer
 	  resource instances.  These are displayed via the
@@ -169,8 +168,7 @@ config LWM2M_CONN_MON_BEARER_MAX
 config LWM2M_CONN_MON_APN_MAX
 	int "Maximum # of APN resource instances"
 	depends on LWM2M_CONN_MON_OBJ_SUPPORT
-	default 5
-	range 1 20
+	default 1
 	help
 	  This value sets the maximum number of APN resource instances.
 	  These are displayed via the "Connection Monitoring" object /4/0/7.
diff --git a/subsys/net/lib/lwm2m/ipso_accelerometer.c b/subsys/net/lib/lwm2m/ipso_accelerometer.c
index 5043588aa3..fcb44e846c 100644
--- a/subsys/net/lib/lwm2m/ipso_accelerometer.c
+++ b/subsys/net/lib/lwm2m/ipso_accelerometer.c
@@ -90,8 +90,10 @@ static struct lwm2m_engine_obj_inst *accel_create(u16_t obj_inst_id)
 	}
 
 	/* Set default values */
-	memset(&accel_data[avail], 0, sizeof(struct ipso_accel_data));
+	(void)memset(&accel_data[avail], 0, sizeof(accel_data[avail]));
 
+	(void)memset(res[avail], 0,
+		     sizeof(res[avail][0]) * ARRAY_SIZE(res[avail]));
 	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
 
 	/* initialize instance resource data */
@@ -125,11 +127,6 @@ static struct lwm2m_engine_obj_inst *accel_create(u16_t obj_inst_id)
 
 static int ipso_accel_init(struct device *dev)
 {
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			ARRAY_SIZE(inst) * ACCEL_MAX_ID);
-
 	accel.obj_id = IPSO_OBJECT_ACCELEROMETER_ID;
 	accel.fields = fields;
 	accel.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/ipso_buzzer.c b/subsys/net/lib/lwm2m/ipso_buzzer.c
index 3e6bf031b9..a3fc528c0a 100644
--- a/subsys/net/lib/lwm2m/ipso_buzzer.c
+++ b/subsys/net/lib/lwm2m/ipso_buzzer.c
@@ -201,18 +201,14 @@ static struct lwm2m_engine_obj_inst *buzzer_create(u16_t obj_inst_id)
 	}
 
 	/* Set default values */
+	(void)memset(&buzzer_data[avail], 0, sizeof(buzzer_data[avail]));
 	k_delayed_work_init(&buzzer_data[avail].buzzer_work, buzzer_work_cb);
 	buzzer_data[avail].level.val1 = 50; /* 50% */
-	buzzer_data[avail].level.val2 = 0;
 	buzzer_data[avail].delay_duration.val1 = 1; /* 1 seconds */
-	buzzer_data[avail].delay_duration.val2 = 0;
-	buzzer_data[avail].min_off_time.val1 = 0;
-	buzzer_data[avail].min_off_time.val2 = 0;
-	buzzer_data[avail].trigger_offset = 0U;
-	buzzer_data[avail].onoff = false;
-	buzzer_data[avail].active = false;
 	buzzer_data[avail].obj_inst_id = obj_inst_id;
 
+	(void)memset(res[avail], 0,
+		     sizeof(res[avail][0]) * ARRAY_SIZE(res[avail]));
 	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
 
 	/* initialize instance resource data */
@@ -249,11 +245,6 @@ static struct lwm2m_engine_obj_inst *buzzer_create(u16_t obj_inst_id)
 
 static int ipso_buzzer_init(struct device *dev)
 {
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			ARRAY_SIZE(inst) * BUZZER_MAX_ID);
-
 	buzzer.obj_id = IPSO_OBJECT_BUZZER_ID;
 	buzzer.fields = fields;
 	buzzer.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/ipso_light_control.c b/subsys/net/lib/lwm2m/ipso_light_control.c
index 048125d7bb..442f82b6ce 100644
--- a/subsys/net/lib/lwm2m/ipso_light_control.c
+++ b/subsys/net/lib/lwm2m/ipso_light_control.c
@@ -161,6 +161,8 @@ static struct lwm2m_engine_obj_inst *light_control_create(u16_t obj_inst_id)
 	colour[avail][0] = '\0';
 	units[avail][0] = '\0';
 
+	(void)memset(res[avail], 0,
+		     sizeof(res[avail][0]) * ARRAY_SIZE(res[avail]));
 	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
 
 	/* initialize instance resource data */
@@ -199,11 +201,6 @@ static struct lwm2m_engine_obj_inst *light_control_create(u16_t obj_inst_id)
 
 static int ipso_light_control_init(struct device *dev)
 {
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * MAX_INSTANCE_COUNT);
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			MAX_INSTANCE_COUNT * LIGHT_MAX_ID);
-
 	light_control.obj_id = IPSO_OBJECT_LIGHT_CONTROL_ID;
 	light_control.fields = fields;
 	light_control.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/ipso_onoff_switch.c b/subsys/net/lib/lwm2m/ipso_onoff_switch.c
index 06a91dc760..47985e7b44 100644
--- a/subsys/net/lib/lwm2m/ipso_onoff_switch.c
+++ b/subsys/net/lib/lwm2m/ipso_onoff_switch.c
@@ -188,14 +188,11 @@ static struct lwm2m_engine_obj_inst *switch_create(u16_t obj_inst_id)
 	}
 
 	/* Set default values */
-	switch_data[avail].trigger_offset = 0U;
-	switch_data[avail].on_time_sec = 0U;
-	switch_data[avail].off_time_sec = 0U;
-	switch_data[avail].counter = 0U;
-	switch_data[avail].state = false;
-	switch_data[avail].last_state = false;
+	(void)memset(&switch_data[avail], 0, sizeof(switch_data[avail]));
 	switch_data[avail].obj_inst_id = obj_inst_id;
 
+	(void)memset(res[avail], 0,
+		     sizeof(res[avail][0]) * ARRAY_SIZE(res[avail]));
 	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
 
 	/* initialize instance resource data */
@@ -227,11 +224,6 @@ static struct lwm2m_engine_obj_inst *switch_create(u16_t obj_inst_id)
 
 static int ipso_switch_init(struct device *dev)
 {
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			ARRAY_SIZE(inst) * SWITCH_MAX_ID);
-
 	onoff_switch.obj_id = IPSO_OBJECT_ONOFF_SWITCH_ID;
 	onoff_switch.fields = fields;
 	onoff_switch.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/ipso_push_button.c b/subsys/net/lib/lwm2m/ipso_push_button.c
index eefb55cc5b..c84ae6985f 100644
--- a/subsys/net/lib/lwm2m/ipso_push_button.c
+++ b/subsys/net/lib/lwm2m/ipso_push_button.c
@@ -120,11 +120,11 @@ static struct lwm2m_engine_obj_inst *button_create(u16_t obj_inst_id)
 	}
 
 	/* Set default values */
-	button_data[avail].counter = 0U;
-	button_data[avail].state = false;
-	button_data[avail].last_state = false;
+	(void)memset(&button_data[avail], 0, sizeof(button_data[avail]));
 	button_data[avail].obj_inst_id = obj_inst_id;
 
+	(void)memset(res[avail], 0,
+		     sizeof(res[avail][0]) * ARRAY_SIZE(res[avail]));
 	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
 
 	/* initialize instance resource data */
@@ -150,11 +150,6 @@ static struct lwm2m_engine_obj_inst *button_create(u16_t obj_inst_id)
 
 static int ipso_button_init(struct device *dev)
 {
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * ARRAY_SIZE(inst));
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			ARRAY_SIZE(inst) * BUTTON_MAX_ID);
-
 	onoff_switch.obj_id = IPSO_OBJECT_PUSH_BUTTON_ID;
 	onoff_switch.fields = fields;
 	onoff_switch.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/ipso_temp_sensor.c b/subsys/net/lib/lwm2m/ipso_temp_sensor.c
index 8947ee3be1..2f3df6ba17 100644
--- a/subsys/net/lib/lwm2m/ipso_temp_sensor.c
+++ b/subsys/net/lib/lwm2m/ipso_temp_sensor.c
@@ -182,6 +182,8 @@ static struct lwm2m_engine_obj_inst *temp_sensor_create(u16_t obj_inst_id)
 	max_range_value[index].val1 = 0;
 	max_range_value[index].val2 = 0;
 
+	(void)memset(res[index], 0,
+		     sizeof(res[index][0]) * ARRAY_SIZE(res[index]));
 	init_res_instance(res_inst[index], ARRAY_SIZE(res_inst[index]));
 
 	/* initialize instance resource data */
@@ -214,11 +216,6 @@ static struct lwm2m_engine_obj_inst *temp_sensor_create(u16_t obj_inst_id)
 
 static int ipso_temp_sensor_init(struct device *dev)
 {
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * MAX_INSTANCE_COUNT);
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			MAX_INSTANCE_COUNT * TEMP_MAX_ID);
-
 	temp_sensor.obj_id = IPSO_OBJECT_TEMP_SENSOR_ID;
 	temp_sensor.fields = fields;
 	temp_sensor.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/ipso_timer.c b/subsys/net/lib/lwm2m/ipso_timer.c
index dc8208ccb5..14f5a277e1 100644
--- a/subsys/net/lib/lwm2m/ipso_timer.c
+++ b/subsys/net/lib/lwm2m/ipso_timer.c
@@ -324,21 +324,15 @@ static struct lwm2m_engine_obj_inst *timer_create(u16_t obj_inst_id)
 	}
 
 	/* Set default values */
+	(void)memset(&timer_data[avail], 0, sizeof(timer_data[avail]));
 	k_delayed_work_init(&timer_data[avail].timer_work, timer_work_cb);
-	timer_data[avail].trigger_offset = 0U;
 	timer_data[avail].delay_duration.val1 = 5; /* 5 seconds */
-	timer_data[avail].delay_duration.val2 = 0;
-	timer_data[avail].remaining_time.val1 = 0;
-	timer_data[avail].remaining_time.val2 = 0;
-	timer_data[avail].min_off_time.val1 = 0;
-	timer_data[avail].min_off_time.val2 = 0;
-	timer_data[avail].cumulative_time.val1 = 0;
-	timer_data[avail].cumulative_time.val2 = 0;
-	timer_data[avail].active = false;
 	timer_data[avail].enabled = true;
 	timer_data[avail].timer_mode = TIMER_MODE_ONE_SHOT;
 	timer_data[avail].obj_inst_id = obj_inst_id;
 
+	(void)memset(res[avail], 0,
+		     sizeof(res[avail][0]) * ARRAY_SIZE(res[avail]));
 	init_res_instance(res_inst[avail], ARRAY_SIZE(res_inst[avail]));
 
 	/* initialize instance resource data */
@@ -390,11 +384,6 @@ static struct lwm2m_engine_obj_inst *timer_create(u16_t obj_inst_id)
 
 static int ipso_timer_init(struct device *dev)
 {
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * MAX_INSTANCE_COUNT);
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			MAX_INSTANCE_COUNT * TIMER_MAX_ID);
-
 	timer.obj_id = IPSO_OBJECT_TIMER_ID;
 	timer.fields = fields;
 	timer.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/lwm2m_obj_connmon.c b/subsys/net/lib/lwm2m/lwm2m_obj_connmon.c
index 1817c8ad19..2dab2aef14 100644
--- a/subsys/net/lib/lwm2m/lwm2m_obj_connmon.c
+++ b/subsys/net/lib/lwm2m/lwm2m_obj_connmon.c
@@ -35,17 +35,8 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
 #define CONNMON_STRING_SHORT			8
 
-#if defined(CONFIG_LWM2M_CONN_MON_BEARER_MAX)
 #define CONNMON_AVAIL_BEARER_MAX	CONFIG_LWM2M_CONN_MON_BEARER_MAX
-#else
-#define CONNMON_AVAIL_BEARER_MAX	5
-#endif
-
-#if defined(CONFIG_LWM2M_CONN_MON_APN_MAX)
 #define CONNMON_APN_MAX			CONFIG_LWM2M_CONN_MON_APN_MAX
-#else
-#define CONNMON_APN_MAX			5
-#endif
 
 #if defined(CONFIG_NET_IF_MAX_IPV6_COUNT) \
 	&& defined(CONFIG_NET_IF_MAX_IPV6_COUNT)
@@ -111,6 +102,13 @@ static struct lwm2m_engine_obj_inst *connmon_create(u16_t obj_inst_id)
 {
 	int i = 0, j = 0;
 
+	/* Set default values */
+	net_bearer = 42U; /* Ethernet */
+	rss = 0;
+	link_quality = 0U;
+	mnc = 0U;
+	mcc = 0U;
+
 	init_res_instance(res_inst, ARRAY_SIZE(res_inst));
 
 	/* initialize instance resource data */
@@ -145,13 +143,6 @@ static int lwm2m_connmon_init(struct device *dev)
 	struct lwm2m_engine_obj_inst *obj_inst = NULL;
 	int ret = 0;
 
-	/* Set default values */
-	net_bearer = 42U; /* Ethernet */
-	rss = 0;
-	link_quality = 0U;
-	mnc = 0U;
-	mcc = 0U;
-
 	/* initialize the Connection Monitoring field data */
 	connmon.obj_id = LWM2M_OBJECT_CONNECTIVITY_MONITORING_ID;
 	connmon.fields = fields;
diff --git a/subsys/net/lib/lwm2m/lwm2m_obj_location.c b/subsys/net/lib/lwm2m/lwm2m_obj_location.c
index b07b1bad6e..e4f5e70522 100644
--- a/subsys/net/lib/lwm2m/lwm2m_obj_location.c
+++ b/subsys/net/lib/lwm2m/lwm2m_obj_location.c
@@ -95,10 +95,6 @@ static int ipso_location_init(struct device *dev)
 	int ret;
 	struct lwm2m_engine_obj_inst *obj_inst = NULL;
 
-	/* Set default values */
-	(void)memset(&inst, 0, sizeof(inst));
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) * LOCATION_MAX_ID);
-
 	location.obj_id = LWM2M_OBJECT_LOCATION_ID;
 	location.fields = fields;
 	location.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/lwm2m_obj_security.c b/subsys/net/lib/lwm2m/lwm2m_obj_security.c
index 0d2fa0bfee..514bca02cd 100644
--- a/subsys/net/lib/lwm2m/lwm2m_obj_security.c
+++ b/subsys/net/lib/lwm2m/lwm2m_obj_security.c
@@ -109,6 +109,8 @@ static struct lwm2m_engine_obj_inst *security_create(u16_t obj_inst_id)
 	security_mode[index] = 0U;
 	short_server_id[index] = 0U;
 
+	(void)memset(res[index], 0,
+		     sizeof(res[index][0]) * ARRAY_SIZE(res[index]));
 	init_res_instance(res_inst[index], ARRAY_SIZE(res_inst[index]));
 
 	/* initialize instance resource data */
@@ -173,11 +175,6 @@ static int lwm2m_security_init(struct device *dev)
 	struct lwm2m_engine_obj_inst *obj_inst = NULL;
 	int ret = 0;
 
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * MAX_INSTANCE_COUNT);
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			MAX_INSTANCE_COUNT * SECURITY_MAX_ID);
-
 	security.obj_id = LWM2M_OBJECT_SECURITY_ID;
 	security.fields = fields;
 	security.field_count = ARRAY_SIZE(fields);
diff --git a/subsys/net/lib/lwm2m/lwm2m_obj_server.c b/subsys/net/lib/lwm2m/lwm2m_obj_server.c
index c5fa5dde0a..f3daec305a 100644
--- a/subsys/net/lib/lwm2m/lwm2m_obj_server.c
+++ b/subsys/net/lib/lwm2m/lwm2m_obj_server.c
@@ -168,6 +168,8 @@ static struct lwm2m_engine_obj_inst *server_create(u16_t obj_inst_id)
 	disabled_timeout[index] = 86400U;
 	strcpy(transport_binding[index], "U");
 
+	(void)memset(res[index], 0,
+		     sizeof(res[index][0]) * ARRAY_SIZE(res[index]));
 	init_res_instance(res_inst[index], ARRAY_SIZE(res_inst[index]));
 
 	/* initialize instance resource data */
@@ -212,11 +214,6 @@ static int lwm2m_server_init(struct device *dev)
 	struct lwm2m_engine_obj_inst *obj_inst = NULL;
 	int ret = 0;
 
-	/* Set default values */
-	(void)memset(inst, 0, sizeof(*inst) * MAX_INSTANCE_COUNT);
-	(void)memset(res, 0, sizeof(struct lwm2m_engine_res) *
-			MAX_INSTANCE_COUNT * SERVER_MAX_ID);
-
 	server.obj_id = LWM2M_OBJECT_SERVER_ID;
 	server.fields = fields;
 	server.field_count = ARRAY_SIZE(fields);
```

jukkar reviewed (approved):
LGTM

rlubos reviewed (approved):
Thanks, LGTM
